Add Muon Optimizer to `contrib`

leloykun opened the pull request:
Adds support for @KellerJordan's Muon optimizer as detailed here: https://kellerjordan.github.io/posts/muon/

This optimizer allowed us to reduce the training time of a GPT-2 level model down to just ~3.5 minutes on 8xH100s. See: https://github.com/KellerJordan/modded-nanogpt

![image](https://github.com/user-attachments/assets/945137da-3c7e-42a2-bb9e-d10abb241d07)

This implementation also supports the adaptive version of Muon that can adapt to the scale of the gradients as they change during training. To enable this feature, simply set `muon(adaptive=True, ...)`. See more at: https://github.com/leloykun/adaptive-muon

---

Muon is equivalent to Shampoo without accumulation. I.e., take Shampoo's update rule, discard the left and right preconditioners, then simplify the matrix operations. Alternatively, Muon can also be thought of as steepest descent under Spectral norm. For more details, please refer to @jxbz's paper https://arxiv.org/pdf/2409.20325.

More generally, steepest descent under Schatten-p norm, for some large p, can be thought of as variants of Muon (note that Schatten-infty norm = Spectral norm). A (rough) proof of this claim can be found here: https://x.com/leloykun/status/1846842887839125941. This implementation supports these variants by allowing the user to pass in custom coefficients for the newton-schulz iterations.

leloykun commented:
Hi all!

How do I restrict the tests to exclude vectors as weights?

vroulet commented:
Hello @leloykun,

Looks like an interesting contribution thank you!

> How do I restrict the tests to exclude vectors as weights?

I don't fully understand your question. Could you give more context?

Other questions/remarks:

How does this optimizer treat vector-shaped values? The"muon_iterator" could run on vectors but not return what you want, so how do you make the distinction? Should it take a mask to only apply on matrices? Should it raise value error?

Also could you add the mathematical description in the doc string? That would greatly help.

Finally, put references at the end of the docstring (we'll adopt this format with #1129)

Thank you!

leloykun commented:
Hi @vroulet,

Muon is only defined for matrix-shaped values.

I'm thinking of raising an error when the input is vector-shaped, but where's the best place to put it? If there are other optimizers here that does this, can you point me to them?

vroulet commented:
Hello @leloykun 

- I would put a test in the init function. 
- What is the algorithm for? Most architectures have non-matrix shaped parameters (for example just the scalar parameters of a layer norm). If the algorithm is specific to some problem, could you make a quick notebook explaining the setup? (Is it for example for finetuning with LoRa?)
- In any case, some doctest would be good.
- For the tests, you may add a synthetic linear classification problem with centered data to test the algorithm for example (again if you have a specific setup let me know).

Thanks again and sorry for the delay!

vroulet commented:
Hey @leloykun,

Looking at the original repository, it seems that the best course of action is to:
- keep the scale_by_muon as you did
- make muon as a [partition](https://github.com/google-deepmind/optax/blob/63cdeb4ada95498626a52d209a029210ac066aa1/optax/transforms/_combining.py#L155#L267) instance where the partition is inferred from the dimensions of the leaves in the optimizer
- add in the tests a simple mlp and make all optimizers pass it

Please let me know if you are willing to pursue the PR :)

leloykun commented:
Hi @rdyro! Thank you so much for the help. This PR should now be ready for review. I've also already addressed @vroulet's concerns. Please let me know if there is anything else I need to do on my end.

rdyro commented on optax/contrib/_muon.py:
```diff
@@ -0,0 +1,254 @@
+"""
+
+
+from typing import Any, NamedTuple, Optional, Tuple, Union
```
nit: could you remove `Optional, Tuple, Union` and use python 3.9+ annotations, e.g., Union[A, B] === A | B, Optional[A] === A | None

rdyro commented on optax/contrib/_muon.py:
```diff
@@ -0,0 +1,254 @@
+
+def scale_by_muon(
+    ns_coeffs: Union[
+      Tuple[float, float, float],
```
nit: could we change Union to the | operator and Tuple to tuple

rdyro commented on optax/contrib/_muon.py:
```diff
@@ -0,0 +1,254 @@
+
+def muon(
+    learning_rate: base.ScalarOrSchedule,
+    ns_coeffs: Union[
```
nit: | and tuple

rdyro commented on optax/contrib/_muon.py:
```diff
@@ -0,0 +1,254 @@
+
+
+def orthogonalize_via_newton_schulz(
+  x: jnp.ndarray,
```
nit: could we use jax.Array instead of jnp.ndarray

rdyro commented on optax/contrib/_muon.py:
```diff
@@ -0,0 +1,254 @@
+      'Newton-Schulz coefficients must have shape (3,) or (n, 3), '
+      f'got {ns_coeffs.shape}'
+    )
+  def newton_schulz_iterator(x: jnp.ndarray, coeffs: jnp.ndarray):
```
nit: could we use jax.Array instead of jnp.ndarray

rdyro reviewed (commented):
Can you introduce these cosmetic changes? The PR looks really good!

rdyro commented:
> Hi @rdyro! Thank you so much for the help. This PR should now be ready for review. I've also already addressed @vroulet's concerns. Please let me know if there is anything else I need to do on my end.

Hey, I just reviewed the PR, it looks great, thanks for addressing @vroulet's comments. I left some cosmetic comments.

Once you get those in, can you squash your commits please?

rdyro commented:
I'm sorry, I mislead you. Because we're supporting Python 3.9 Union and Tuple are necessary, exactly like you had it (having Optional would be ok too, but you can leave it as Union[A, None]). I'll start the merge once you can revert those Union/Tuple changes.

leloykun commented:
@rdyro, I've just addressed your nits and squashed the commits.

I kept the `Union` & `Optional` type hints tho as support for `|` was only added in v3.10 instead of v3.9... See: https://peps.python.org/pep-0604/

jkyl commented:
Thank you for this contribution, I think the way that you partitioned the matrix params from the non-matrix ones is totally reasonable. I did just want to point out an important edge case which is not handled here yet. [When using `vmap` to create a stack of layers](https://flax.readthedocs.io/en/latest/nnx_basics.html#scan-over-layers), the layer stack's weight matrices will be stacked along the zeroth axis, creating a 3-dimensional array. This kind of param won't get picked up by Muon in this implementation. It's not obvious to me how this should be handled -- maybe its best to expose the partitioning condition to the user somehow?

stephen-huan commented:
Also convolution filters, I think? cf. [Muon](https://github.com/KellerJordan/Muon/blob/12dc5336852ec10fbe1a1119ad09c92d2d81505e/muon.py#L97-L98)

```python
                    if g.ndim == 4: # for the case of conv filters
                        g = g.view(len(g), -1)
```
